gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/writer/FileAwareInputStreamDataWriterBuilder.java

```diff
@@ -52,14 +53,13 @@
   public synchronized static void setJobSpecificOutputPaths(State state) {
 
     // Other tasks may have set this already
+    // If reading from the sharded dataset path, writer directory paths are stored in the workunit state
     if (!StringUtils.containsIgnoreCase(state.getProp(ConfigurationKeys.WRITER_STAGING_DIR),
-        state.getProp(ConfigurationKeys.JOB_ID_KEY))) {
-
-      state.setProp(ConfigurationKeys.WRITER_STAGING_DIR, new Path(state.getProp(ConfigurationKeys.WRITER_STAGING_DIR),
-          state.getProp(ConfigurationKeys.JOB_ID_KEY)));
-      state.setProp(ConfigurationKeys.WRITER_OUTPUT_DIR, new Path(state.getProp(ConfigurationKeys.WRITER_OUTPUT_DIR),
-          state.getProp(ConfigurationKeys.JOB_ID_KEY)));
-
+        state.getProp(ConfigurationKeys.JOB_ID_KEY)) && !state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR)) {
+        state.setProp(ConfigurationKeys.WRITER_STAGING_DIR, new Path(state.getProp(ConfigurationKeys.WRITER_STAGING_DIR),
```

Comment:
Can all of this be moved inside the handler?

Reply:
Same comment as earlier. We should be able to set override the WRITER_STAGING_DIR config inside the DatasetHandler and leave the code here unchanged.

Reply:
@Will-Lo  If your concern is about hive-distcp user with shard configuration: It should be fine to move this into handler only since I don't think there's any real usage existed for shard-aware distcp